File: Day92/Part6/smart_economy.py
import sys
import math
from collections import deque

def solve():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
        
    iterator = iter(input_data)
    try:
        N = int(next(iterator))
        M = int(next(iterator))
    except StopIteration:
        return

    edges = []
    adj = [[] for _ in range(N + 1)]
    
    for _ in range(M):
        u = int(next(iterator))
        v = int(next(iterator))
        r = float(next(iterator))
        edges.append((u, v, r))
        adj[u].append(v)
        
    # Bellman-Ford Maximize Log Sum
    # dist[u] = max log accumulated
    dist = [-float('inf')] * (N + 1)
    dist[1] = 0.0
    
    # Relax N-1 times
    for _ in range(N - 1):
        updated = False
        for u, v, r in edges:
            if dist[u] > -1e15: # Reachable
                w = math.log(r) if r > 0 else -float('inf')
                if dist[u] + w > dist[v] + 1e-9:
                    dist[v] = dist[u] + w
                    updated = True
        if not updated:
            break
            
    # Check cycles
    # Nodes that can still be relaxed are part of positive cycle flow
    in_cycle_flow = [False] * (N + 1)
    queue = deque()
    
    for u, v, r in edges:
        if dist[u] > -1e15:
            w = math.log(r) if r > 0 else -float('inf')
            if dist[u] + w > dist[v] + 1e-9:
                # Cycle detected affecting v
                if not in_cycle_flow[v]:
                    in_cycle_flow[v] = True
                    queue.append(v)
    
    # Propagate cycle flow
    while queue:
        u = queue.popleft()
        for v in adj[u]:
            if not in_cycle_flow[v]:
                in_cycle_flow[v] = True
                queue.append(v)
                
    if in_cycle_flow[N]:
        print("INFINITY")
    else:
        if dist[N] < -1e15:
            print("0.00000")
        else:
            print(f"{math.exp(dist[N]):.5f}")

if __name__ == "__main__":
    solve()

File: Day92/Part6/smart_economy.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>
#include <queue>
#include <iomanip>

using namespace std;

// Problem 6: Economy With Turing-Complete Contracts
// (Interpreted as: Currency Exchange with Arbitrage Detection)
// Maximize wealth at target node N, given start node 1 with 1.0 unit.
// Edges u -> v with rate w.
// Use Bellman-Ford with Logarithms to find longest path and detect positive cycles.

struct Edge {
    int u, v;
    double rate;
};

const double INF = 1e18; // Logic infinity for distances

int main() {
    int N, M;
    if (!(cin >> N >> M)) return 0;

    vector<Edge> edges;
    vector<vector<int>> adj(N + 1); // For reachability check
    
    for (int i = 0; i < M; ++i) {
        int u, v;
        double w;
        cin >> u >> v >> w;
        edges.push_back({u, v, w});
        adj[u].push_back(v);
    }
    
    // Bellman-Ford
    // Use log probabilities.
    // rate R -> log(R).
    // Multiply rates -> Add logs.
    // Maximize sum of logs.
    // Init dists to -INF.
    
    vector<double> dist(N + 1, -INF);
    dist[1] = 0.0; // log(1) = 0
    
    // Run N-1 iterations
    for (int i = 0; i < N - 1; ++i) {
        bool updated = false;
        for (const auto& e : edges) {
            if (dist[e.u] > -INF * 0.5) { // If reachable
                double new_dist = dist[e.u] + log(e.rate);
                if (new_dist > dist[e.v] + 1e-9) {
                    dist[e.v] = new_dist;
                    updated = true;
                }
            }
        }
        if (!updated) break;
    }
    
    // Check for positive cycles (which appear as relaxable edges after N-1 steps)
    // And check if such a cycle can reach Node N.
    
    vector<bool> in_cycle_flow(N + 1, false);
    queue<int> q;
    
    for (const auto& e : edges) {
        if (dist[e.u] > -INF * 0.5) {
             double new_dist = dist[e.u] + log(e.rate);
             if (new_dist > dist[e.v] + 1e-9) {
                 // e.v is updated, so it is part of or reachable from a positive cycle
                 // We need to see if this propagates to N
                 if (!in_cycle_flow[e.v]) {
                     in_cycle_flow[e.v] = true;
                     q.push(e.v);
                 }
                 // Logic: Just pushing e.v to a queue and doing BFS to see what it hits.
             }
        }
    }
    
    while(!q.empty()){
        int u = q.front();
        q.pop();
        
        for(int v : adj[u]){
            if(!in_cycle_flow[v]){
                in_cycle_flow[v] = true;
                q.push(v);
            }
        }
    }
    
    if (in_cycle_flow[N]) {
        cout << "INFINITY" << endl;
    } else {
        if (dist[N] <= -INF * 0.5) {
            cout << "0.00000" << endl; // Unreachable
        } else {
            cout << fixed << setprecision(5) << exp(dist[N]) << endl;
        }
    }

    return 0;
}
